### PR TITLE
C++/winRT migration

### PR DESCRIPTION
**Pull request by brabebhin**
Hi all,

This is the PR for migrating the library to C++/winRT. The migration currently exists as a separate project, with a separate namespace, but this is just nicities for later.

What happened here, in a nuttshell:
1. All classes that were not needed to be winRT classes have been moved to vanila C++ (mostly sample providers and other low level stuff that integrates with FFMPEG).
2. There is still some work to be done on the hardware accelerated video effect (haven't migrated that but it should be mostly mechanical work).
3. Need to fix all bugs and test this well.

Some additional breaking changes:
1. Metadata tags are now returned as Map<string, Vector<string>> instead of Vector<KeyValuePair<string, string>>. Mostly a small difference.
What do you think, @lukasf  @JunielKatarn ?

### CODE REVIEW

**Comment by brabebhin** [comment]
Am I dumb or is it really hard to implement the IBuufferByteAccess in c++/winrt?

**Review by JunielKatarn — COMMENTED** [review]
(Note: still reviewing).

General notes:
- Rename FFmpegInteropxWinUI to FFmpegInteropX.WinUI
  Also for namespaces, prefer nesting: FFmpegInterop::WinUI
- Consider using standard C++ instead of C++/CX.

File: FFmpegInteropXWinUI/AbstractEffectFactory.h
Comment (brabebhin) [reply FFmpegInteropXWinUI/AbstractEffectFactory.h:9]: Prevents direct class instanctation.

File: FFmpegInteropXWinUI/AttachedFile.h
Comment (brabebhin) [reply FFmpegInteropXWinUI/AttachedFile.h:8]: These are auto generated files, I assume they are properly generated.

File: FFmpegInteropXWinUI/AttachedFile.h
Comment (brabebhin) [reply FFmpegInteropXWinUI/AttachedFile.h:32]: These "AVxxxxxx" pointers are ffmpeg pointers which is a C library, and they also use some custom memory cleanup. Rather not mix them.

File: FFmpegInteropXWinUI/AttachedFileHelper.h
Comment (brabebhin) [reply FFmpegInteropXWinUI/AttachedFileHelper.h:5]: Because the ffmpeg library is a C library, the extern "C" things tells the linker to use C linkage.

File: FFmpegInteropXWinUI/AttachedFileHelper.h
Comment (brabebhin) [reply FFmpegInteropXWinUI/AttachedFileHelper.h:18]: Fair point, not sure why this didn't trigger a compiler error. lol.

File: FFmpegInteropXWinUI/AbstractEffectFactory.h
Comment (brabebhin) [reply FFmpegInteropXWinUI/AbstractEffectFactory.h:9]: ah now I realize this also is a leftover from c++/CX. Again, not sure why the compiler isn't issuing an error, I use the "C++/winRT windows runtime component tenplate" from visual studio to generate that project,

Goes to show just how much of a freak show this c++/winRT actually is. As I struggle to port stuff I wonder if this is really worth it and it isn't just a temporary tantrum from Microsoft.

**Comment by brabebhin** [comment]
I am quite close to getting it 100% working. There's a problem with the texture and buffer pools, objects returned in the pool can't be recycled apparently.

**Comment by brabebhin** [comment]
So directx and software decoding seems to work for video now. I will need to fix passthroughs.

Most bugs seem to originate from uninitialized variables.

**Comment by lukasf** [comment]
Wow @brabebhin, really nice work! This is really amazing. Kudos for all the heavy lifting.

I did not yet have time for a deep review. But some general remarks:

- When the migration is complete, this should be a drop-in replacement. So the namespace should be the same and also no API changes, if possible (the metadata API change is accepted). I guess you used a separate namespace for development and testing, which is fine. But finally, we should use the old namespace. From the consumer perspective, it should not make a difference if our Windows Runtime Component was developed using C++/CX or C++/WinRT.

- The files are in a new folder, which will make git lose the history. If you are more or less finished with migration, the new files should be put back into the old Source folder (any leftover files deleted there). Then I can review old against new implementation, which will be much better then just looking at the new sources.

- There are a lot of changesets in here. I think we should make this a squash commit, to get a clean history. We need to be extra careful when merging this PR.

- We also need to think about when we will merge this. Since the changes are so massive, we probably should not hurry this one out. Maybe we should first push out the original 1.0 release PR (I planned to do this before my long journey, but didn't have the time in the end). Then we can thoroughly test this and make it a 1.1, once we are confident there are no leaks or bugs. We can include the metadata API change into the 1.0 release already, so we won't have API changes between 1.0 and 1.1.

**Comment by brabebhin** [comment]
@lukasf agreed. We need serious testing on this.

I originally wanted to do drop in migration, but the project file got all mingled up and decided starting fresh is just easier, and once everything works, I just rename the namespace and move the files back in.

**Comment by lukasf** [comment]
Sure, if you like, you can push the API changes to 1.0 branch as well.

I can do the NuGet stuff once we are ready for a release.

**Comment by brabebhin** [comment]
@JunielKatarn 

In your experience, does it make sense to initialize hstring instances as empty strings? like this? hstring bla = L"";?

File: FFmpegInteropXWinUI/AttachedFileHelper.h
Comment (JunielKatarn) [reply FFmpegInteropXWinUI/AttachedFileHelper.h]: Likely, CL.exe still supports C++/CX constructs.
I don't know if there is an explicit way to "turn them off".

File: FFmpegInteropXWinUI/AbstractEffectFactory.h
Comment (JunielKatarn) [reply FFmpegInteropXWinUI/AbstractEffectFactory.h:9]: C++/WinRT is not a language dialect, nor a language extension.
It's a set of standard C++ generated projections for lower level Windows and COM APIs.

In my opinion, it's very worth it because of its standards compliance.
It also saves a lot of boilerplate code as opposed to using plain WINAPI.

**Comment by JunielKatarn** [comment]
> @JunielKatarn
> 
> In your experience, does it make sense to initialize hstring instances as empty strings? like this? hstring bla = L"";?

I believe this is the "preferred modern C++" initialization style: `winrt::hstring myHstring{};`.

**Comment by lukasf** [comment]
So I merged the v1.0 PR. This one will need a rebase.

@brabebhin Do you have an account on nuget.org, by the way? I'd like to make you second owner for the libs, as a backup.

**Comment by brabebhin** [comment]
Hi,

Just created one now, same username, "Brabebhin".

**Comment by lukasf** [comment]
Okay, done. Updating the NuGet libs is pretty simple. Just click on „Upload“ on top, drop the created .nupkg, verify that the displayed metadata looks fine (version number, etc), and then proceed. It may take a few minutes until the libs go live.

I'll take care of this normally, but it's probably good if you can do it as well, just in case.

**Comment by brabebhin** [comment]
Got the invites, thanks.
I hope you're good :)

**Comment by lukasf** [comment]
Yes, the vacation was pretty amazing, thanks! :)

**Comment by lukasf** [comment]
On a short test, the port seems to run fine on my PC. Basic playback working and no major memory leaks after opening a bunch of files. Fixed some NuGet issues and aligned toolset versions, trying to get AppVeyor build working...

**Comment by brabebhin** [comment]
I will need to port the BasicVideoEffect these coming days.

**Comment by brabebhin** [comment]
So apparently using win2D is quite the hustle. We need to download a specific version of win2D, use cppwinrt.exe to generate some header files and pack the dll along with the winmd. This will needs to be done manually...

**Comment by lukasf** [comment]
The effect is not buggy as hell, I just broke it with the 1.0 namespace change.

On StackOverflow, I read that with more recent Win2D versions (as of mid 2018), you just need to reference the NuGet package and then the headers will be auto-generated. Did you try that?

Also Microsoft is moving Win2D into the WindowsAppSdk, and they have a separate NuGet package for WinUI 3 apps (which is not 100% finished as I read). But if we use that, it probably won't work in UWP apps. So if we cannot get the normal Win2D package working, we might have to move the effect out to a separate addon library, specific for WinUI 3. Maybe it would be a good idea anyways, to have a separate lib for stuff like the video effect and custom subtitle rendering, which are both more or less independend of the FFmpegInteropX core library.

**Comment by brabebhin** [comment]
I wasn't referring to the activation when i said it's buggy. For some reason, simply installing the nuget doesn't work for me. But if it works does you,  then it means there's something wrong with me setup.

**Comment by brabebhin** [comment]
Thanks for the hint @lukasf 

After installing some new updates to visual studio, it does generate the header files for me automatically. Which is great, I pushed the changeset.

All i need to do now is understand how the activatable class id for the effect works in C++/winRT. Apparently it is not as easy as in C++/CX. I'll leave this for later.

**Comment by brabebhin** [comment]
Turns out the problem was unrelated to the class ID, which is the same as in C++/CX, but rather was another idl/ cppwinrt gotcha:

The IDL of the class must contain a parameterless constructor. The [default_interface] attribute does not suffice, the constructor must be declared explicitly.

**Comment by lukasf** [comment]
I just pushed a big changeset to address all those compiler warnings. Some of them actually pointed to problems and even bugs. The basic funtionality seems to be working now. Maybe it is a good time to move the files back, so we can start a proper review of all file changes.

Of course, this will need intense testing of all the special cases and functionalities, with various files and formats.

**Comment by brabebhin** [comment]
I actually think we should merge the other outstanding PR before migrating the namespace and folder.

**Comment by brabebhin** [comment]
Hi @lukasf 

I merged the branch fix-leaks. Can you check and see if the merge was ok?

Thanks.

**Comment by lukasf** [comment]
Looks good. I done some minor changes there.

We should also consider making use of com_ptr class in WinRT for all the D3D stuff. It really simplifies the ref handling, making it more solid at the same time. And you can just use `.as<T>` instead of QueryInterface, which is also a lot nicer.

**Comment by brabebhin** [comment]
Yes. That is a good idea. I also thought of it, but i initially decided to port the code as is and do stuff like this later.

On the other hand, since there's no rush, i could do it before merge.

**Comment by lukasf** [comment]
We can do it later as well. The merging will probably be easier, if we do not do too much of refactoring upfront.

**Comment by lukasf** [comment]
By the way, in the WinRT samples, I often see assignments through blocks like this:

```
int main()
{
    winrt::Windows::Foundation::IStringable s{ winrt::make<MyStringable>() };
    auto result{ s.ToString() };
}
```

Do you have any insights on why they do it like that? Can't we just write:
```
int main()
{
    winrt::Windows::Foundation::IStringable s = winrt::make<MyStringable>();
    auto result = s.ToString();
}
```

Their samples are full of that assignment style, but they never explain why it is done like that. I wonder if it has any benefits.

**Comment by brabebhin** [comment]
As per my understanding, it depends on what you initialize. For a class with constructor, it basically calls the constructor. For classes without constructors, it initializes the members of the class in order.  This is quite the gotcha in some directx12 structures.  I lost days debugging something like that. In your examples i think it doesn't make any difference between the two examples. 

In some situations it also helps the compiler disambiguate situations in which you have an variable with the same name as a function and your try to call the constructor of that variable.

**Comment by lukasf** [comment]
Yeah I only know this style from initialization from structs. And it comes with the disadvantage that you don't know which parameter goes into which field, if you don't know the exact struct layout.

But if you have a constructor, then why not call it directly? Why not store the result of a method with `=`? To me it feels like abusing a language construct that was ment for something different. And without any obvious benefit, at least I don't see it. I really wonder who came up with that idea, and why they decided to use it so heavily.

**Comment by lukasf** [comment]
I just read some interesting things about strong and weak references in C++/WinRT. In the old implementation, we had this requirement that we need to store the FFmpegMediaSource reference in the app, to prevent it beging collected during playback. With C++/WinRT, we can add get_strong() to the MSS event registration code. Then the MSS would hold a strong reference to the FFmpegMediaSource and prevent any early collection. No more need to store it in app code.

https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/weak-references#safely-accessing-the-this-pointer-with-an-event-handling-delegate

Also in other event handlers, we should think about lifetime issues, and consider using strong or weak references explicitly, where appropriate.

**Comment by brabebhin** [comment]
Interesting find. I think that would basically solve the problem if we register events in the mss (like the starting event) with our ffmpeg media source as a strong reference.

I think we need to decide of we want to do those refactorings now. It could be useful to do all possible dangerous changes in one go.

Perhaps something along the lines of:

Merge stream buffer
Put files back in folder
Review changes
Do dangerous stuff(com ptr, references etc) 
Review again
Merge

**Comment by lukasf** [comment]
I agree that it is a good idea to test all this together. But I would really like to keep the history clean, and not mix too many things into one PR. So I would propose to bring all these to master, but as separate PRs (winrt, stream buffer, refactorings). Each PR should be roughly tested and reviewed. But final testing will be done on master branch once all things are there, before we make it an official release.

I think that it will also make reviews better, if you can see a clean diff of e.g. only the pure winrt migration without refactoring, only stream buffer, only com_ptr refactoring,...

What do you think?

**Comment by brabebhin** [comment]
This is also a good idea. But we need to decide  if we want to merge this before stream buffer or after.

Once this is merged I will also import it in my own app and i will be able to do some more advanced testing on it.

**Comment by lukasf** [comment]
To me it does not really matter now which one gets merged first. Whatever feels finished first gets merged first. There will be some manual merge effort when these two get in touch, but I think it does not really matter in which order it arrives.

I am still not 100% happy with the stream buffer implementation. It works, but it has some design weakness that I don't like. I had to use workarounds to avoid circular references. Maybe I will come up with a better design to avoid these workarounds and have cleaner code. So that branch might also need some more time.

We should probably do a 1.0.1 release with the fix for the BasicVideoEffect, before anything else gets merged to master.

**Comment by brabebhin** [comment]
Ok, then i will put the files back in this week. Probably today or tomorrow, if things go well.

**Comment by brabebhin** [comment]
OK, I've moved the files back, but the build isn't working. There seems to be a problem with the precompiled header which I can't figure out, tried several approached and all ended in the same problem. I pushed this so maybe you can also take a look and see what I'm missing. I will keep trying.

**Comment by lukasf** [comment]
I fixed most of the build but two errors remain, regarding WinRT actication factory.
Probably need to compare project files more.

**Comment by brabebhin** [comment]
Cool. I will focus on that now. Nice job with the pch, I tried various combinations and none worked :( (I even thought its contents were to blame somehow, hence commenting it out)

File: Source/FFmpegInteropX.vcxproj
Comment (brabebhin) [inline Source/FFmpegInteropX.vcxproj:6]: I've been staring at this for 2 hours, I don't understand, what was the problem??

Comment (lukasf) [reply Source/FFmpegInteropX.vcxproj:6]: I really don't know why, but it looked like the position of the third import statements was what was breaking things. After I moved it below the "Globals" section (where it was originally), things were back working.

**Comment by lukasf** [comment]
The C# sample works fine, but I found two issues with the CPP sample, surprisingly:

1. The binding of CharacterEncodings does not work. We can get the value find, but an ArgumentException comes when assigning it to Combobox.ItemsSource.
2. When opening a file, somehow the app continues on a wrong thread, causing an exception when accessing the splitter (close pane). co_await the file open returns normally on UI thread, but co_await on our MediaStreamSource somehow returns on a background thread. That should not happen of course. Pretty weird. Playback still works, but there is something wrong with the async stuff we return in our CreateFromXX methods.

We need to find out what is wrong here.

**Comment by lukasf** [comment]
I fixed the second issue by manually switching back to UI thread. To me, this does not seem right. Normally, the co_await on the UI page should switch us back to UI thread, no matter what the lib does. But it seems that the WinRT resume on background switches the context in a way that UI won't switch back. This is very different from what I know async/await. I have kind of a bad feeling about this.

I tried a few things about the CharacterEncoding but failed. The old class has that Bindable attribute. But the attribute is missing in the winrt headers. And if I include the legacy header, I get loads of compile errors. I have also seen [bindable] in the idl file in some docs, so I added that but it didn't help as well. I am not even sure if that is the problem.

**Comment by brabebhin** [comment]
The problem seems to be the VectorView that we return rather than the CharacterEncoding class.

**Comment by lukasf** [comment]
It's strange that the observable vector works, but the normal vector does not work. The normal vector is supposed to work as well, and it does in C++/CX. It's just that it won't report updates, but we are not updating this one anyways.

I needed to add the [bindable] attribute in the IDL file, otherwise binding to properties of CharacterEncoding did not work for the CPP sample. We should consider adding this to any class that might be bound in a UI, such as the StreamInfo classes.

And I just found two problems in the subtitle providers:

Looks like style parsing does not work anymore on the C++/WinRT branch. sscanf_s always returns 0 parameters parsed, although the input data looks perfectly fine. On master, the same input data returns 23 parameters parsed.

And also the code to parse the actual subtitle lines seems to be one index off, but that occurs on master as well?? I only tested with one file, need to test more. It might be a change in FFmpeg. The ssa text is not in the standard format anymore, looks like the first parameter is missing. I wonder how long this is broken. Subtitles are still displayed, but styles are not applied anymore.

**Comment by lukasf** [comment]
I tested this in an actual app and found some more issues, especially subtitle related. There was also a crash on cleanup, and some other things. Now it is looking pretty good, still more testing needed.

I tried a few hacks to get XML docs working, but it only works for methods, not properties, events or enums. It's a shame that C++/WinRT does not even support XML docs. Such a basic thing not working, every other language has that.

I noticed that AttachedFile is part of the IDL. It is only used internally, we should keep that as a pure C class. The current AttachedFile would break when our main class is destroyed. If we'd want to expose AttachedFiles, we'd need to think about refernce handling.

**Comment by brabebhin** [comment]
AttachedFile is public in the current c++/cx version.

**Comment by brabebhin** [comment]
Sorry I was out and about the past few days. I see what you mean with the AttachedFile, however, that api has been public for quite some time. Changing it would be a breaking change.

**Comment by lukasf** [comment]
That's true. But since there was no way to create it (no public ctor, no other API), I would not really consider this a breaking change. I have created a PR to remove it from old implementation as well.

**Comment by brabebhin** [comment]
TBH I never used that API since we merged it the first time around.

File: Source/UncompressedVideoSampleProvider.cpp
Comment (lukasf) [inline Source/UncompressedVideoSampleProvider.cpp:619]: Check if this buffer is released properly. Not related to WinRT migration.

File: Source/BasicVideoEffect.h
Comment (brabebhin) [reply Source/BasicVideoEffect.h:55]: I avoided using namespace directives cause in some cases it would cause ambiguity problems.

File: Source/CharacterEncoding.cpp
Comment (brabebhin) [reply Source/CharacterEncoding.cpp:21]: True.

File: Source/DirectXInteropHelper.cpp
Comment (brabebhin) [reply Source/DirectXInteropHelper.cpp:19]: We cannot use the helper function from C++/CX, they are simply not available to c++/winrt. This directx interop file was one of the biggest pain points to migrate.

File: Source/FrameGrabber.cpp
Comment (brabebhin) [reply Source/FrameGrabber.cpp:14]: Your problem might stem from mixing C++/CX with C++/winRT.

File: Source/MediaSampleProvider.h
Comment (brabebhin) [reply Source/MediaSampleProvider.h:180]: All members that were once "internal" are now public. This is because C++/winRT does not support internal members. I don't see how we could achieve the same effect.

Comment (lukasf) [reply Source/MediaSampleProvider.h:180]: My point is that it was wrong to have them internal, already in the old implementation. We should not directly access fields from outside the class. Fields should be private or protected, but not internal or public.

File: Source/BasicVideoEffect.h
Comment (lukasf) [reply Source/BasicVideoEffect.h:55]: The only place where we seem need them is when accessing FFmpegInteropX classes from within FFmpegInteropX::implementation class. For all other namespaces, usings work fine. At least, I just cleaned up namespaces on the FFmpegMediaSource and saw no issues with that (except as I said, for direct FFmpegInteropX type access).

Comment (brabebhin) [reply Source/BasicVideoEffect.h:55]: We also use some WRL headers and that causes problems because both c++/winRT and WRL define IInspectable.

File: Source/MediaSampleProvider.h
Comment (brabebhin) [reply Source/MediaSampleProvider.h:180]: For fields which are both read and write, we shouldn't encapsulate, it would be just noise.
I think we inherited many internal stuff from the original repo.

File: Source/BasicVideoEffect.h
Comment (lukasf) [reply Source/BasicVideoEffect.h:55]: There are a few cases where collisions occur. IUnknown is one other interface that is defined twice. But that should not lead us to not have namespace usings at all. That makes code very verbose. It is always possible to fix collisions by using explicit namespaces where required.

File: Source/UncompressedVideoSampleProvider.h
Comment (brabebhin) [reply Source/UncompressedVideoSampleProvider.h:108]: Can be null at some point.

File: Source/DirectXInteropHelper.cpp
Comment (brabebhin) [reply Source/DirectXInteropHelper.cpp]: The fact that documentation links to the wrong pages doesn't help either....

File: Source/BasicVideoEffect.h
Comment (brabebhin) [reply Source/BasicVideoEffect.h:55]: Well, I guess we can work out backwards. But using any single using namespace winrt::blablabla includes IInspectable and IUnknown. But I don't think we should do it now.

Comment (lukasf) [reply Source/BasicVideoEffect.h:55]: It's not a problem. Just add explicit namespace where you use IInspectable or IUnknown. Then there is no collision. I will cleanup some of the bigger classes before merging this.

Comment (brabebhin) [reply Source/BasicVideoEffect.h:55]: I actually tried that approach and got tangled in the spaghetiness of microsoft header libraries. Eventually I gave up and simply fully qualified almost everywhere. 

It might be that you're right and maybe my errors were actually ghost errors that the C++ compiler likes to produce, and now that we have a clean state we can work back and remove some of the full qualifiers. 

We should also consider removing the old "ffmpeginteropx" namespace that contains the pure c++ classes, and leave those outside a namespace. This standalone namespace sometimes conflicts with winrt::fffmpeginteropx

Comment (lukasf) [reply Source/BasicVideoEffect.h]: I also thought about removing the namespace. Even the ::implementation classes can be removed from namespace. It does not matter where they are or how they are called. They just need to be there with the right template parameters. Microsoft does not use any namespaces for their implementation classes in WinUI 3.

File: Source/UncompressedVideoSampleProvider.h
Comment (lukasf) [reply Source/UncompressedVideoSampleProvider.h:108]: Oh right. Thanks for clearing that up.

File: Source/UncompressedVideoSampleProvider.cpp
Comment (lukasf) [reply Source/UncompressedVideoSampleProvider.cpp:619]: Okay this is all fine.

File: Source/BasicVideoEffect.h
Comment (brabebhin) [reply Source/BasicVideoEffect.h]: I agree with removing the ::implementation namespace if it helps. That would greatly simplify our casts between winRT and implementation types.

File: Source/SubtitleProviderSsaAss.h
Comment (lukasf) [reply Source/SubtitleProviderSsaAss.h:570]: Okay this was only moved from way above. Should be fine.

File: Source/BasicVideoEffect.h
Comment (lukasf) [reply Source/BasicVideoEffect.h]: I already fixed up the bigger classes now. Removing ::implementation is a bigger task. I think we can do that later (unless you like to do it now).

Comment (brabebhin) [reply Source/BasicVideoEffect.h]: I can give it a try during the weekend. Unless you want to merge this now.

Comment (lukasf) [reply Source/BasicVideoEffect.h]: I am not in a hurry. Depends on how fast the other issues are closed. I definitely want to take a closer look at the event handler thing. C++/WinRT is different than C++/CX in that area, if I understood correctly. I want to be sure that we get the lifetime things correct.

File: Source/SubtitleProvider.h
Comment (lukasf) [reply Source/SubtitleProvider.h]: Fixed

File: Source/SubtitleProviderSsaAss.h
Comment (lukasf) [reply Source/SubtitleProviderSsaAss.h]: This is also fixed now. I think that after upping platform min version to 10.0.17763.0, we could probably drop a lot of ApiInformation checks. But that can also be done later.

File: Source/FrameGrabber.cpp
Comment (lukasf) [reply Source/FrameGrabber.cpp:16]: That could be the case. Anyways, we could do this later, as it's working fine. Closing for now.

**Comment by brabebhin** [comment]
Nice work! Surely trumps my template method in pch. 

If i remember correctly, i only used one event handler with weak references, can't remember which one

File: Source/AttachedFileHelper.h
Comment (lukasf) [reply Source/AttachedFileHelper.h:96]: Okay so std::map does not work because StorageFile does not have default ctor. Yeah I read something about that in the docs. C++/WinRT really has its pitfalls, using structs over classes everywhere.

**Comment by lukasf** [comment]
For me this is looking all good now. Maybe a little final testing and we are good to merge.

File: Source/SubtitleProviderSsaAss.h
Comment (brabebhin) [reply Source/SubtitleProviderSsaAss.h:571]: I think this was a winRT gotcha with pass by value vs pass by reference.

**Comment by brabebhin** [comment]
Cool. I will give it some short tests over the weekend.

On a side note: biggest PR yet?

**Comment by brabebhin** [comment]
Looking good for me.

**Review by lukasf — APPROVED** [review]

**Comment by lukasf** [comment]
We are on master now. Nice work!

**Comment by MouriNaruto** [comment]
Really good to see this work.

Kenji Mouri

**Pull request merged**
